src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -184,7 +184,15 @@ private function parseDefinition(\DOMElement $service, $file)
                 } elseif ($childService = $factory->getAttribute('service')) {
                     $class = new Reference($childService, ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE, false);
                 } else {
-                    $class = $factory->getAttribute('class');
+                    $class = $factory->getAttribute('class') ?: $definition->getClass();
+                }
+
+                if (null === $class) {
```

Comment:
Not sure it's the loader responsibility.
Also the class might not be available at this point, in case it's generated.